backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/DedicatedCpuPinningHelper.java

```diff
@@ -0,0 +1,212 @@
+
+        pinPendingExclusiveCpus(cpuTopology, vmToPendingPinnings, vm.getCpuPinningPolicy());
+        List<VdsCpuUnit> cpusToBeAllocated = new ArrayList<>();
+
+        if (vm.getCpuPinningPolicy() != CpuPinningPolicy.DEDICATED) {
+            String cpuPinning = vm.getCpuPinningPolicy() == CpuPinningPolicy.MANUAL ? vm.getCpuPinning() : vm.getCurrentCpuPinning();
+            Set<Integer> requestedCpus = CpuPinningHelper.getAllPinnedPCpus(cpuPinning);
+            for (Integer cpuId : requestedCpus) {
+                VdsCpuUnit vdsCpuUnit = getCpu(cpuTopology, cpuId);
+                vdsCpuUnit.setCpuPinningPolicy(vm.getCpuPinningPolicy());
+                vdsCpuUnit.setVmId(vm.getId());
+                cpusToBeAllocated.add(vdsCpuUnit);
+            }
+            return cpusToBeAllocated;
+        }
+
+
+        int cpus = vm.getNumOfCpus();
+        int cpusLeft = cpus;
+        int hostCoresPerSocket = host.getCpuCores() / host.getCpuSockets();
+        int hostThreadsPerCore = host.getCpuThreads() / host.getCpuCores();
+
+        for (int socket = 0; socket < host.getCpuSockets(); socket++) {
+            if (hostCoresPerSocket * hostThreadsPerCore >= cpus && getFreeCpusInSocket(cpuTopology, socket).size() < cpus) {
+                // we should fit into one socket
+                continue;
+            }
+            cpusLeft = allocateCores(cpuTopology, hostCoresPerSocket, hostThreadsPerCore, socket, cpusLeft,
+                    cpusToBeAllocated, true, vm.getId(), vm.getCpuPinningPolicy());
+            if (cpusLeft > 0 && getFreeCpusInSocket(cpuTopology, socket).size() >= cpusLeft) {
+                // iterate again on the cores, take whatever we can.
+                cpusLeft = allocateCores(cpuTopology, hostCoresPerSocket, hostThreadsPerCore, socket, cpusLeft,
+                        cpusToBeAllocated, false, vm.getId(), vm.getCpuPinningPolicy());
+            }
+        }
+        return cpusLeft == 0 ? cpusToBeAllocated : new ArrayList<>();
+    }
+
+    private int allocateCores(List<VdsCpuUnit> cpuTopology, int hostCoresPerSocket, int hostThreadsPerCore, int socket,
+                              int cpusLeft, List<VdsCpuUnit> cpusToBeAllocated, boolean wholeCore, Guid vmId,
+                              CpuPinningPolicy cpuPinningPolicy) {
+        for (int core = 0; core < hostCoresPerSocket; core++) {
+            List<VdsCpuUnit> sharedCpus = getFreeCpusInCore(cpuTopology, socket, core);
+            int sharedCpusInCore = sharedCpus.size();
+            List<Integer> sharedCpuIds = sharedCpus.stream().map(VdsCpuUnit::getCpu).collect(Collectors.toCollection(LinkedList::new));
+            Iterator<Integer> sharedCpusIterator = sharedCpuIds.iterator();
+            if (cpusLeft == 0) {
+                return cpusLeft;
+            }
+            if (wholeCore) {
+                if (cpusLeft >= hostThreadsPerCore && sharedCpusInCore == hostThreadsPerCore) {
```

Comment:
Why `cpusLeft >= hostThreadsPerCore`?

Reply:
As I wish to take the whole core, I want to be sure I have at least the number of vCPUs as there are in this core. Otherwise I will take only partial core.

On the second loop, if there is no partial core that fits, we will get to the same place and take it. But we do wish to first get only whole cores and we can't get it, do partial.